gameobject.h

```diff
@@ -43,6 +43,11 @@ class GameObject
     bool CheckForCollision(GameObject& other);
 
     void Render() const;
+    template <typename T>
+    T* GetObject()
+    {
+        return dynamic_cast<T*>(this);
```

Comment:
This function leaves me feeling a bit funny. First off, because it's at compile time I think you could possibly get away with static_cast. Second, in what situation would you need to get an object in another type?

Reply:
With functions like OnCollisionStart() you get a GameObject&. However if when handling a collision you need something specific from the other object you would need to get it into that type first. A lot of this comes from how unity handles it. For example, in Unity this is a typical instance of handling a collision.
```
public void OnTriggerEnter(Collider collider)
{
    if (collider.tag == "projectile")
    {
        // Get the actual object out from the collider.
        Projectile projectile = collider.GetComponent<Projectile>();
        // could check here to make sure it actually was an instance of Projectile (projectile != null)
        TakeDamage(projectile.GetDamage());
     }
}